content/en/blog/2022/debug-otel-with-otel/index.md

```diff
@@ -0,0 +1,259 @@
+
+### What did you see instead?
+
+In the jaeger UI at [localhost:16686][] you will see two traces, one going from
+the `frontend` down to nginx, and another one only for the `python-app`.
+
+## The solution
+
+### The hints
+
+Since the setup worked with a java application in the backend, we knew that the
+problem was either caused by the python application or by the combination of the
+nginx instrumentation and the python application.
+
+We could quickly rule out that the python application alone was the issue:
+trying out a simple Node.JS application as backend, we got the same result: two
+traces, one from frontend to nginx, another one for the Node.JS application
+alone.
+
+With that, we knew that we had a propagation issue: the trace context was not
+transferred successfully from nginx down to the python and Node.JS application.
+
+### The analysis
+
+Knowing that the issue does not occur with java and that it is likely a broken
+propagation, we knew what we had to do: we needed to see the trace headers.
+
+Gladly, the instrumentations for [Java][] and [Python][] have a feature that allows us
+to capture [HTTP request & response headers][] as span attributes easily.
+
+By providing a comma-separated list of HTTP header names via the environment
+variables `OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST` and
+`OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_RESPONSE` we can define which
+HTTP headers we want to capture. In our case we put all potential propagation
+headers:
+
+```console
+OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST=tracestate,traceparent,baggage,X-B3-TraceId
+```
+
+In our `docker-compose`-based example we simply can add it to the definition of
+our backend service:
+
+```yaml
+  backend:
+    build: ./backend
+    image: backend-with-otel
+    environment:
+      - OTEL_EXPORTER_OTLP_ENDPOINT=http://collector:4318/v1/traces
+      - OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
+      - OTEL_SERVICE_NAME=python-app
+      - OTEL_INSTRUMENTATION_HTTP_CAPTURE_HEADERS_SERVER_REQUEST=tracestate,traceparent,baggage,X-B3-TraceId
+```
+
+Once again we ran `docker-compose up` to bring up our sample app and we send
+some request with `curl localhost:8080` to the frontend application.
+
+In Jaeger we still see that the traces are disconnected. However, when we look
+into one of those traces, we can see the collected request headers from nginx to
+backend:
+
+![A screenshot of the Jaeger UI showing that http.request.header.traceparent has multiple entries.](jaeger-with-request-headers.png)
+
+There it is! The trace headers (`baggage`, `traceparent`, `tracestate`) are send
+as multiple header fields.
+
+Doing some research on that topic, we found out, that this is coved by
+[RFC7230][] and is [allowed and defined][] for `tracestate`, but the [definition
+for `traceparent`][] does not provide details on that corner-case.
+
+Although we didn't read into the source code of the OTel Java SDK & OTel Python
```

Comment:
Should this manifest in issues for otel-py and otel-js?

Reply:
I don't know what the expected behavior is, since the w3c spec does not define that.